### PR TITLE
feat: added flags for event audit

### PR DESCRIPTION
**Pull request by kanishkkatara**
# Description

Fetching event audit enabled flag from workspace config and using it in processor to send events to esch_jobs table

## Linear Ticket

https://linear.app/rudderstack/issue/DAT-285/add-ability-for-customer-to-self-serve-the-enablement-disablement-of

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: processor/processor.go [hunk 1]
```diff
@@ -731,13 +731,15 @@ func (proc *Handle) backendConfigSubscriber(ctx context.Context) {
 				}
 			}
 			workspaceLibrariesMap[workspaceID] = wConfig.Libraries
+			eventAuditEnabled = wConfig.Settings.EventAuditEnabled
```
Comment (abhimanyubabbar) [inline processor/processor.go:734]: You would override the value in this loop as it is running for multiple workspaces right ?

**Review by abhimanyubabbar — CHANGES_REQUESTED** [review]

**Review by atzoum — CHANGES_REQUESTED** [review]
`eventAuditEnabled` needs to be a `map[string]bool` protected by `configSubscriberLock`

File: processor/processor.go [hunk 2]
```diff
@@ -1490,8 +1492,7 @@ func (proc *Handle) processJobsForDest(partition string, subJobs subJob) *transf
 			},
 			)
 			if proc.config.eventSchemaV2Enabled && // schemas enabled
-				// source has schemas enabled or if we override schemas for all sources
-				(source.EventSchemasEnabled || proc.config.eventSchemaV2AllSources) &&
```
Comment (atzoum) [inline processor/processor.go:1494]: Please delete `SourceT#EventSchemasEnabled` in `backendconfig/types.go` and adapt `processor_test.go`

File: processor/processor.go [hunk 3]
```diff
@@ -725,7 +724,9 @@ func (proc *Handle) backendConfigSubscriber(ctx context.Context) {
 			sourceIdDestinationMap = make(map[string][]backendconfig.DestinationT)
 			sourceIdSourceMap      = map[string]backendconfig.SourceT{}
 			destinationIDtoTypeMap = make(map[string]string)
+			eventAuditEnabled      = make(map[string]bool)
 		)
+		proc.config.configSubscriberLock.Lock()
```
Comment (atzoum) [inline processor/processor.go:729]: please revert this change, the lock shall be obtains as before, to minimize contention

File: processor/processor.go [hunk 4]
```diff
@@ -1490,8 +1492,7 @@ func (proc *Handle) processJobsForDest(partition string, subJobs subJob) *transf
 			},
 			)
 			if proc.config.eventSchemaV2Enabled && // schemas enabled
-				// source has schemas enabled or if we override schemas for all sources
-				(source.EventSchemasEnabled || proc.config.eventSchemaV2AllSources) &&
+				proc.config.eventAuditEnabled[batchEvent.WorkspaceId] &&
```
Comment (atzoum) [inline processor/processor.go:1495]: You need to acquire a read lock before reading the value from `eventAuditEnabled` here

File: processor/processor.go [hunk 5]
```diff
@@ -1490,8 +1498,8 @@ func (proc *Handle) processJobsForDest(partition string, subJobs subJob) *transf
 			},
 			)
 			if proc.config.eventSchemaV2Enabled && // schemas enabled
-				// source has schemas enabled or if we override schemas for all sources
-				(source.EventSchemasEnabled || proc.config.eventSchemaV2AllSources) &&
+				proc.eventAuditEnabled(batchEvent.WorkspaceId) &&
+				proc.config.eventAuditEnabled[batchEvent.WorkspaceId] &&
```
Comment (atzoum) [inline processor/processor.go:1502]: oops
```suggestion
```

**Review by atzoum — APPROVED** [review]

File: backend-config/types.go [hunk 6]
```diff
@@ -121,7 +120,8 @@ func (c *ConfigT) DestinationsMap() map[string]*DestinationT {
 }
 
 type Settings struct {
-	DataRetention DataRetention `json:"dataRetention"`
+	DataRetention     DataRetention `json:"dataRetention"`
+	EventAuditEnabled bool          `json:"eventAuditEnabled"`
```
Comment (abhimanyubabbar) [inline backend-config/types.go:124]: Have we verified that data is unmarshalling into this struct automatically ?

File: backend-config/backend-config.go [hunk 7]
```diff
@@ -156,6 +156,8 @@ func filterProcessorEnabledDestinations(config ConfigT) ConfigT {
 		source.Destinations = destinations
 		modifiedConfig.Sources = append(modifiedConfig.Sources, source)
 	}
+	modifiedConfig.Settings.DataRetention.DisableReportingPII = config.Settings.DataRetention.DisableReportingPII
```
Comment (atzoum) [inline backend-config/backend-config.go:159]: can't we just use the whole `Settings` struct instead of selectively picking fields??

Comment (kanishkkatara) [reply backend-config/backend-config.go:159]: DataRetention has other properties that will get overridden, thats why didn't do this

Comment (atzoum) [reply backend-config/backend-config.go:156]: Not including workspace settings in the `TopicProcessConfig` looks like a bug to me, cc @cisse21, @BonapartePC

File: backend-config/types.go (hunk 6 above)
Comment (kanishkkatara) [reply backend-config/types.go:127]: yes

**Review by lvrach — APPROVED** [review]

File: backend-config/backend-config.go (hunk 7 above)
Comment (cisse21) [reply backend-config/backend-config.go]: Yeah true.... this looks like a bug which we had.

**Review by abhimanyubabbar — APPROVED** [review]
